### ISSUE TITLE
Add floats to AST, write comprehensive parser tests for them

### ISSUE THREAD

**Issue by Janiczek** [elm-in-elm/compiler#17]
This means:

- [ ] add Float constructor to the Literal type... solve any compiler errors that arise as a result (emitExpr etc - look at the Int cases for inspiration, it should be pretty similar)
- [x] find all the ways floats can be written in Elm (`1.2`? `1.`? scientific notation like `2e10`? maybe stuff like `NaN`, `Infinity` etc? (probably not))... go see official Elm's parser to be sure!
- [ ] add those as tests as a "literal float" section of ParserTest.elm (probably doesn't exist yet) so that the tests exercise all the possibilities and truthfully tell us what's done and what's not...
- [ ] add emit tests too
- [ ] update the "what's done" table in README - parser tests for floats, but probably also optimizations (none), desugaring, typechecking and emitting

**Comment by Janiczek** [comment]
Seems like all the ways are:
* `1.2` (positive float)
* `-1.2` (negative float)
* `5.2e3` (scientific notation)
* `-5.2e3` (negative number before exponent)
* `5.2E3` (big E)
* `5e-2` (negative exponent)

**Comment by aaronjanse** [comment]
Looks like adding it to the AST shouldn't be too difficult using [`Parser.float`](https://package.elm-lang.org/packages/elm/parser/latest/Parser#float) :+1:

**Comment by Janiczek** [comment]
That might very well be the case, I'm just wary of the `elm/parser` bugs wrt. Parser.int and Parser.float:
https://github.com/elm/parser/issues

Anyways, even if we found out we can't use these functions and had to roll our own, doing what the reference implementation does shouldn't be _too_ hard: https://github.com/elm/compiler/blob/master/compiler/src/Parse/Number.hs

**Comment by Warry** [comment]
The `number` parser seems adapted:

```
literalNumber : Parser_ Literal
literalNumber =
    let
        parseLitteralNumber =
            P.number
                { int = Ok Int
                , hex = Ok Int
                , octal = Ok Int
                , binary = Ok Int
                , float = Ok Float
                , invalid = InvalidNumber
                , expecting = ExpectingNumber 
                }

        negateLitteral toBeNegated =
            case toBeNegated of
                Int int ->
                    Int (negate int)

                Float float ->
                    Float (negate float)

                _ ->
                    toBeNegated

    in
    P.oneOf
        [ P.succeed negateLitteral
            |. P.symbol (P.Token "-" ExpectingMinusSign)
            |= parseLitteralNumber
        , parseLitteralNumber
        ]
        |> P.inContext InLiteralNumber
```

**Comment by Janiczek** [comment]
@Warry Can you create a PR? I think we can test against any `elm/parser` bugs using our test suite.

**Comment by Warry** [comment]
I'll do it during lunch break :)

**Comment by Janiczek** [comment]
fixed by 0a4ffb7
